clippy_lints/src/vec.rs

```diff
@@ -46,16 +51,70 @@ declare_clippy_lint! {
 
 impl_lint_pass!(UselessVec => [USELESS_VEC]);
 
+fn adjusts_to_slice(cx: &LateContext<'_>, e: &Expr<'_>) -> bool {
+    if let ty::Ref(_, ty, _) = cx.typeck_results().expr_ty_adjusted(e).kind() {
+        ty.is_slice()
+    } else {
+        false
+    }
+}
+
+/// Checks if the given expression is a method call to a `Vec` method
+/// that also exists on slices. If this returns true, it means that
+/// this expression does not actually require a `Vec` and could just work with an array.
+fn is_allowed_vec_method(cx: &LateContext<'_>, e: &Expr<'_>) -> bool {
+    const ALLOWED_METHOD_NAMES: &[&str] = &["len", "as_ptr", "is_empty"];
```

Comment:
I'm a bit curious why `len` and `is_empty` are allowed.

Reply:
The method names that are checked for here exist on both slices and `Vec<T>`.
So, given a variable `ve` of type `Vec<i32>`, a call expression like `ve.is_empty()` does not actually call `slice::is_empty`, but `Vec::is_empty`, which also means that `adjusts_to_slice(ve)` would return false.

If we didn't have `len` and `is_empty` here, the lint would assume that those are `Vec<T>` methods that really do require a `Vec<T>` (like `push` and `pop`, for example).
For `len` and `is_empty` specifically, that isn't the case. These two methods also exist on slices, and we still want to lint for cases like these, so we allow those:
```rs
let ve = vec![1, 2];
dbg!(ve.is_empty()); // Vec::is_empty, but this would also work fine if ve was `[1, 2]`
```

Reply:
Ah, I misunderstood then :sweat_smile: I thought those were exceptions. Yeah, that sounds fine then.

Reply:
nit: please add a comment to this const explaining what it's about